### PR TITLE
water particle

### PR DESCRIPTION
**Pull request by tcm390**

### CODE REVIEW

**Review by aryasalehi — CHANGES_REQUESTED** [review]

File: layers/water-mesh.js [hunk 1]
```diff
@@ -264,7 +267,9 @@ export class WaterMesh extends BufferedMesh {
       this.gpuTasks.delete(key);
     }
   }
-
+  initSetUp() {
```
[the remark this answers is not in the document]
Comment (tcm390) [reply layers/water-mesh.js]: [init particle in constructor](https://github.com/webaverse/heightfield/pull/12/commits/7bcd44fb8f419494d931a73f70bf2ea28799e452)

File: layers/water-mesh.js [hunk 2]
```diff
@@ -417,4 +419,20 @@ export class WaterMesh extends BufferedMesh {
...
     }
   }
+  setPackage(pkg) {
+    const particleTextures = pkg.textures['particleTextures'];
+    const particleModels = pkg.models['particleModels'];
+    this.particleEffect = new WaterParticleEffect(particleTextures, particleModels);
+  }
+
+  async waitForLoad() {
+    const baseUrl = import.meta.url.replace(/(\/)[^\/\\]*$/, '$1');
```
Comment (tcm390) [inline layers/water-mesh.js]: removed
[clean code](https://github.com/webaverse/heightfield/pull/12/commits/6f1bf572f0292647f32680ccaeff8c369e38cef9)

File: meshes/water-package.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+        }
+      }
+      return obj;
+    });
+
+    const particleGLBArray = mapObjectToArray(particleGLBPath);
+    const particleModels = await Promise.all(particleGLBArray.map(_loadModel))
+    .then(function(arr) {
+      const obj = {};
+      for (let i= 0; i < particleGLBArray.length; i++) {
```
Comment (tcm390) [inline meshes/water-package.js]: [bad space](https://github.com/webaverse/heightfield/pull/12/commits/403c50a24caf3619b3e64e721ff8934f58e22d45)

File: meshes/water-package.js [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      for (const key in obj)
+        res.push({key: key, value: obj[key]});
+      return res;
+    }
+
+    const particleTextureArray = mapObjectToArray(particleTexturePath);
+    const particleTextures = await Promise.all(particleTextureArray.map(_loadTexture))
+    .then(function(arr) {
+      const obj = {};
+      for (let i= 0; i < particleTextureArray.length; i++) {
```
Comment (tcm390) [inline meshes/water-package.js]: [bad space](https://github.com/webaverse/heightfield/pull/12/commits/403c50a24caf3619b3e64e721ff8934f58e22d45)

File: water-effect/loadTexture.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
+import * as THREE from 'three';
```
[the remark this answers is not in the document]
Comment (tcm390) [reply water-effect/loadTexture.js]: [remove loadTexture.js](https://github.com/webaverse/heightfield/pull/12/commits/fd796bed99c467f1c187d09f8fb91cc727f822e6)

File: water-effect/mesh.js [hunk 6]
```diff
@@ -0,0 +1,538 @@
...
+  bubbleVertex, bubbleFragment,
+  dropletRippleVertex, dropletRippleFragment,
+  movingRippleVertex, movingRippleFragment,
+  freestyleSplashVertex, freestyleSplashFragment,
+  bodyDropVertex, bodyDropFragment,
+} from './shader.js';
+
+const baseUrl = import.meta.url.replace(/(\/)[^\/\\]*$/, '$1');
+const {useLoaders} = metaversefile;
+const voronoiNoiseTexture = loadTexture(`./assets/textures/voronoiNoise.jpg`);
```
[the remark this answers is not in the document]
Comment (tcm390) [reply water-effect/mesh.js]: [separate loading texture and loading glb](https://github.com/webaverse/heightfield/pull/12/commits/1364422f3d9c4c2259878e534b994a92ebb4ca3d)

File: water-effect/particle.js [hunk 7]
```diff
@@ -0,0 +1,1008 @@
...
+        // emit freestyle splash on swimming hand
+        this.playFreestyleSplash();
+      } 
+    }
+    this.lastSwimmingHand = currentSwimmingHand; 
+
+    if (timestamp - this.lastMovingSplashEmitTime > 30 && currentSwimmingHand) {
+      const rightScale = animationType === 'freestyle' 
+                  ? (currentSwimmingHand === 'right' ? 1.2 + Math.random() * 0.2 : 0.8 + Math.random() * 0.2)
+                  : 1.5 * this.currentSpeed + Math.random() * 0.2;
```
[the remark this answers is not in the document]
Comment (tcm390) [reply water-effect/particle.js:634]: [clean code](https://github.com/webaverse/heightfield/pull/12/commits/45a4d5cae9f1c55308f2634ec8d6dcdcb797d7c7)

File: water-effect/particle.js [hunk 8]
```diff
@@ -0,0 +1,1008 @@
...
+
+  handleSwimmingSplash(timestamp, animationType) {
+    if (this.currentSpeed < 0.1) {
+      return;
+    }
+    
+    const currentSwimmingHand = this.player.avatarCharacterSfx.currentSwimmingHand;
+    const playerQ = localVector4.set(this.playerDir.x, this.playerDir.y, this.playerDir.z).applyQuaternion(rotateY);
+    
+    if (currentSwimmingHand && this.lastSwimmingHand !== currentSwimmingHand) {
```
[the remark this answers is not in the document]
Comment (tcm390) [reply water-effect/particle.js:583]: [clean code](https://github.com/webaverse/heightfield/pull/12/commits/45a4d5cae9f1c55308f2634ec8d6dcdcb797d7c7)

File: water-effect/assets.js [hunk 9]
```diff
@@ -0,0 +1,23 @@
+const baseUrl = import.meta.url.replace(/(\/)[^\/\\]*$/, '$1');
+
+export const glbUrlSpecs = {
+  particleGLBPath: {
+    ripple: `${baseUrl}../water-effect/assets/ripple.glb`,
```
Comment (aryasalehi) [inline water-effect/assets.js:5]: all of the assets should be added to this repo :
https://github.com/webaverse/procgen-assets
just like the terrain assets

File: water-effect/particle.js [hunk 10]
```diff
@@ -0,0 +1,1021 @@
...
+
+  playDivingLowerSplash() {
+    const brokenAttribute = this.divingLowerSplash.geometry.getAttribute('broken');
+    const positionsAttribute = this.divingLowerSplash.geometry.getAttribute('positions');
+    const scalesAttribute = this.divingLowerSplash.geometry.getAttribute('scales');
+    const textureRotationAttribute = this.divingLowerSplash.geometry.getAttribute('textureRotation');
+    const particleCount = this.divingLowerSplash.info.particleCount;
+    for (let i = 0; i < particleCount; i ++) {
+      this.divingLowerSplash.info.velocity[i].x = Math.sin(i) * .055 + (Math.random() - 0.5) * 0.001;
+      this.divingLowerSplash.info.velocity[i].y = 0.12 + 0.01 * Math.random();
```
Comment (aryasalehi) [inline water-effect/particle.js:120]: why these values don't have names ?

File: water-effect/particle.js [hunk 11]
```diff
@@ -0,0 +1,1021 @@
...
+
+  playDivingHigherSplash() {
+    const brokenAttribute = this.divingHigherSplash.geometry.getAttribute('broken');
+    const positionsAttribute = this.divingHigherSplash.geometry.getAttribute('positions');
+    const scalesAttribute = this.divingHigherSplash.geometry.getAttribute('scales');
+    const rotationAttribute = this.divingHigherSplash.geometry.getAttribute('rotation');
+    const particleCount = this.divingHigherSplash.info.particleCount;
+    for (let i = 0; i < particleCount; i ++) {
+      this.divingHigherSplash.info.velocity[i].y = 0.08 + 0.035 * Math.random();
+      brokenAttribute.setX(i, 0.2 + Math.random() * 0.1);
```
Comment (aryasalehi) [inline water-effect/particle.js:178]: I don't get the random math, these values should be variables with names

File: water-effect/particle.js [hunk 12]
```diff
@@ -0,0 +1,1021 @@
...
+            positionsAttribute.getX(i) + this.bubble.info.velocity[i].x,
+            positionsAttribute.getY(i) + this.bubble.info.velocity[i].y,
+            positionsAttribute.getZ(i) + this.bubble.info.velocity[i].z
+          );
+
+          // handle texture offset
+          this.bubble.info.livingTime[i] ++;
+          if (this.bubble.info.livingTime[i] % 2 === 0)
+            this.bubble.info.offset[i] ++;
+          if (this.bubble.info.offset[i] >= 30)
```
Comment (aryasalehi) [inline water-effect/particle.js:753]: These values should be variables, constants at the top of the file, so that we can change them easily

**Review by aryasalehi — CHANGES_REQUESTED** [review]
I added some comments for cleaning this up
But other than that, this PR is huge, please break this into smaller PRs ( like 3 ) so that it's easier to review

**Pull request closed without merging**